polkadot/node/core/approval-voting/src/lib.rs

```diff
@@ -3163,7 +3174,50 @@ where
 		}
 		if newly_approved {
 			state.record_no_shows(session_index, para_id.into(), &status.no_show_validators);
+			let mut collected_useful_approvals = vec![];
+
+			match status.required_tranches {
```

Comment:
Could you please extract this logic in a functions, please and add some tests for it.